nyxx_lavalink/lib/src/Cluster.dart

```diff
@@ -0,0 +1,258 @@
+part of nyxx_lavalink;
+
+class _Cluster {
+  /// A reference to the client
+  final Nyxx _client;
+
+  /// The client id provided to lavalink;
+  final Snowflake _clientId;
+
+  /// All available nodes, ordered by node id
+  final Map<int, Node> _nodes = {};
+
+  /// Nodes that are currently connecting to server, when a node gets connected
+  /// it will be moved to [_nodes], and when reconnecting will be moved here again
+  final Map<int, Node> _connectingNodes = {};
+
+  /// A map to keep the assigned node id for each player
+  final Map<Snowflake, int> _nodeLocations = {};
+
+  int _lastId = 0;
+
+  final _receivePort = ReceivePort();
+  late final Stream<dynamic> _receiveStream;
+
+  final _logger = logging.Logger("Lavalink cluster");
+
+  /// Emitted when stats are sent from lavalink
+  late final Stream<Stats> onStatsReceived;
+  /// Emitted when a player gets updated
+  late final Stream<PlayerUpdate> onPlayerUpdate;
+  /// Emitted when a track starts playing
+  late final Stream<TrackStart> onTrackStart;
+  /// Emitted when a track ends playing
+  late final Stream<TrackEnd> onTrackEnd;
+  /// Emitted when a web socket is closed
+  late final Stream<WebSocketClosed> onWebSocketClosed;
+  /// Emitted everytime an event is received
+  late final Stream<Map<String, dynamic>> onRawEvent;
+
+  late final _EventDispatcher _eventDispatcher;
+
+  Future<void> _addNode(NodeOptions nodeOptions, int nodeId) async {
+    await Isolate.spawn(_handleNode, this._receivePort.sendPort);
+
+    final isolateSendPort = await this._receiveStream.firstWhere((element) => element is SendPort) as SendPort;
+
+    nodeOptions.clientId = this._clientId;
+    nodeOptions.nodeId = nodeId;
+
+    isolateSendPort.send(nodeOptions._toJson());
+
+    // Say the node to start the connection
+    isolateSendPort.send({"cmd": "CONNECT"});
+
+    final node = Node._fromOptions(this, nodeOptions, isolateSendPort);
+
+    this._connectingNodes[nodeId] = node;
+  }
```

Comment:
I tried to reproduce the way a shard works on nyxx, also, as there can be as many nodes as the user wants, if those nodes were executing on main isolate, this could slow down the entire application